processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java

```diff
@@ -302,8 +302,13 @@ public boolean equals(Object obj) {
         }
 
         for ( int i = 0; i < getSourceParameters().size(); i++ ) {
-            if ( !getSourceParameters().get( i ).getType().getTypeParameters().get( 0 )
-                .equals( other.getSourceParameters().get( i ).getType().getTypeParameters().get( 0 ) ) ) {
+            List<Type> thisTypeParameters = getSourceParameters().get( i ).getType().getTypeParameters();
+            List<Type> otherTypeParameters = other.getSourceParameters().get( i ).getType().getTypeParameters();
+
+            if ( thisTypeParameters.isEmpty() && otherTypeParameters.isEmpty() ) {
+                continue;
+            }
+            if ( !thisTypeParameters.get( 0 ).equals( otherTypeParameters.get( 0 ) ) ) {
```

Comment:
I think the case that `thisTypeParameters` is empty and `otherTypeParameters` is not empty (and vice versa) is missing. Perhaps you could check on `...size() != ...size()` (as above with the sourceParameters) and then iterate over them to check each parameter pair?